### PR TITLE
Fix cluster cleanup race

### PR DESCRIPTION
**Pull request by cmurphy**
There is a race condition where the cluster object on the local cluster
is being deleted, but the cluster agent on the downstream cluster is not
yet cleaned up, and without this change the controller won't get a
kubeconfig for the deleted cluster and won't be able to create the
cleanup job.

To fix this for RKE1, an exception was made in ToRESTConfig to have it
return a kubeconfig even as the cluster was being deleted, as long as
the driver was RKE1. This doesn't help for the case where the driver is
"Imported" or "RKE2". It's also problematic because it means that
controllers will be unnecessarily started when the cluster is being
deleted.

This change moves the deletion check out of ToRESTConfig and into the
internal controller handling methods, and reworks the imported cluster
cleanup and node cleanup handlers to use clustermanager.ToRESTConfig and
config.NewUserContext, circumventing clustermanager.UserContext which
starts controllers. With the deletion check moved out of ToRESTConfig,
cleanup handlers can safely use it to get credentials for the cluster
and launch cleanup jobs.

https://github.com/rancher/rancher/issues/34486

### CODE REVIEW

**Review by thedadams — APPROVED** [review]
As long as master PR gets merged, this looks good.

**Comment by cmurphy** [comment]
PR to master was closed, this is now the main PR.

See Dan's comment from the original: https://github.com/rancher/rancher/pull/34557#pullrequestreview-743239686

**Comment by snasovich** [comment]
@luthermonson @dramich , just a friendly reminder to review this PR today or tomorrow. Thanks!

File: pkg/clustermanager/manager.go [hunk 1]
```diff
@@ -117,6 +117,9 @@ func (m *Manager) markUnavailable(clusterName string) {
 }
 
 func (m *Manager) start(ctx context.Context, cluster *v3.Cluster, controllers, clusterOwner bool) (*record, error) {
+	if cluster.DeletionTimestamp != nil {
```
Comment (dramich) [inline pkg/clustermanager/manager.go:120]: I don't think we should add this here. `UserContext` is called from a lot of places and so far the only one that seemed to have an issue is for the cleanup which is addressed below. I am concerned this might cause other bugs and affect other flows we aren't aware of or testing.

File: pkg/clustermanager/manager.go [hunk 2]
```diff
@@ -400,6 +397,10 @@ func VerifyIgnoreDNSName(caCertsPEM []byte) (func(rawCerts [][]byte, verifiedCha
 }
 
 func (m *Manager) toRecord(ctx context.Context, cluster *v3.Cluster) (*record, error) {
+	if cluster.DeletionTimestamp != nil {
```
Comment (dramich) [inline pkg/clustermanager/manager.go:400]: Same comment, but this is only called from start so not really needed either way.

File: pkg/controllers/management/node/cleanup.go [hunk 3]
```diff
@@ -36,7 +37,19 @@ func (m *Lifecycle) deleteV1Node(node *v3.Node) (runtime.Object, error) {
...
+	if err != nil {
+		return node, err
+	}
+	if kubeConfig == nil {
+		logrus.Debugf("could not get kubeconfig for cluster %s", cluster.Name)
+		return node, nil
+	}
+	userClient, err := config.NewUserContext(m.clusterManager.ScaledContext, *kubeConfig, cluster.Name)
 	if err != nil {
 		if kerror.IsNotFound(err) {
```
Comment (dramich) [inline pkg/controllers/management/node/cleanup.go:54]: This should be moved up to the lister call

**Review by dramich — APPROVED** [review]

**Review by luthermonson — DISMISSED** [review]

File: pkg/clustermanager/manager.go (hunk 1 above)
Comment (cmurphy) [reply pkg/clustermanager/manager.go:120]: I added this back, I think it's necessary to fix the issue of unnecessarily starting controllers since it was removed from `ToRESTConfig`. This should have the same behavior since before this it would hit the same check in `ToRESTConfig`. Please re-review.

File: pkg/clustermanager/manager.go [hunk 4]
```diff
@@ -468,6 +467,20 @@ func (m *Manager) UserContext(clusterName string) (*config.UserContext, error) {
 	return record.cluster, nil
 }
 
+// UserContextFromCluster accepts a pointer to a Cluster and returns a client
+// for that cluster. It does not start any controllers.
+func (m *Manager) UserContextFromCluster(cluster *v3.Cluster) (*config.UserContext, error) {
```
Comment (snasovich) [inline pkg/clustermanager/manager.go:472]: Kudos for adding documentation comments.
I'm thinking if we could update func name to better convey the fact it doesn't start controllers. There is not much difference between `UserContext` and `UserContextFromCluster`, especially considering both take some info from cluster (either name or pointer to complete Cluster object). `UserContextNoControllers` or something like this?

**Review by dramich — APPROVED** [review]

File: pkg/clustermanager/manager.go [hunk 5]
```diff
@@ -117,6 +117,9 @@ func (m *Manager) markUnavailable(clusterName string) {
 }
 
 func (m *Manager) start(ctx context.Context, cluster *v3.Cluster, controllers, clusterOwner bool) (*record, error) {
+	if cluster.DeletionTimestamp != nil {
+		return nil, nil
+	}
```
Comment (dramich) [inline pkg/clustermanager/manager.go:122]: We should remove this and just revert the change in `ToRESTConfig`, that gets us back to the old logic.

Comment (dramich) [reply pkg/clustermanager/manager.go:122]: So remove the `&& cluster.Status.Driver != v32.ClusterDriverRKE` check and the comment

Comment (cmurphy) [reply pkg/clustermanager/manager.go:122]: The timestamp check is in `ToRESTConfig` is the whole reason the imported cluster cleanup is failing. If I put it back there then I would have to write a new function to get a rest config for deleting clusters.

Comment (dramich) [reply pkg/clustermanager/manager.go:122]: Yup, never mind
